gerrit: Make package definition generic and add other supported releases

felixsinger opened the pull request:
```
The Gerrit project usually maintains the last 3 minor releases and there
might be reasons for not using the most recent one, like a breaking
change or a bug. NixOS has only the most recent version packaged at the
moment.

To give people using the NixOS package and module infrastructure some
more flexibility, allow adding multiple Gerrit packages by making the
package definition generic, which takes the hash and the version as
additional arguments, and move the bits specific to version 3.14 to
`3_14.nix`.

Versioned package names follow the scheme `gerrit_y_xx`,
e.g. `gerrit_3_14`, while `gerrit` points to the most recent version
managed by package.nix.

To make sure `passthru.tests` uses the appropriate package, set
`services.gerrit.package` to `finalAttrs.finalPackage` in the generic
package definition.

Use majorMinor version as base string to be matched in the update
script, so that only point releases are considered. Customize the path
to the Nix file containing the version and hash strings which should be
patched. Otherwise it tries to patch generic.nix.

The following command was used to verify new versions are detected and
patched correctly.

    nix-shell maintainers/scripts/update.nix --argstr package gerrit_3_14
```

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [x] [NixOS tests] in [nixos/tests].
  - [x] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.
- [x] Follows the [automation/AI policy].

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[automation/AI policy]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#automationai-policy
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

felixsinger committed:
gerrit: Make package definition generic

The Gerrit project usually maintains the last 3 minor releases and there
might be reasons for not using the most recent one, like a breaking
change or a bug. NixOS has only the most recent version packaged at the
moment.

To give people using the NixOS package and module infrastructure some
more flexibility, allow adding multiple Gerrit packages by making the
package definition generic, which takes the hash and the version as
additional arguments, and move the bits specific to version 3.14 to
`3_14.nix`.

Versioned package names follow the scheme `gerrit_y_xx`,
e.g. `gerrit_3_14`, while `gerrit` points to the most recent version
managed by package.nix.

To make sure `passthru.tests` uses the appropriate package, set
`services.gerrit.package` to `finalAttrs.finalPackage` in the generic
package definition.

Use majorMinor version as base string to be matched in the update
script, so that only point releases are considered. Customize the path
to the Nix file containing the version and hash strings which should be
patched. Otherwise it tries to patch generic.nix.

The following command was used to verify new versions are detected and
patched correctly.

    nix-shell maintainers/scripts/update.nix --argstr package gerrit_3_14

Signed-off-by: Felix Singer <felixsinger@posteo.net>

felixsinger committed:
gerrit_3_12: init at 3.12.9

Signed-off-by: Felix Singer <felixsinger@posteo.net>

felixsinger commented:
@ofborg build gerrit_3_11 gerrit_3_12 gerrit_3_13

felixsinger committed:
gerrit_3_13: init at 3.13.8

Signed-off-by: Felix Singer <felixsinger@posteo.net>

felixsinger committed:
gerrit: Enable strictDeps and structuredAttrs

Signed-off-by: Felix Singer <felixsinger@posteo.net>

felixsinger commented on pkgs/by-name/ge/gerrit/generic.nix:
```diff
@@ -0,0 +1,69 @@
+      url = "https://gerrit.googlesource.com/gerrit";
+      rev-prefix = "v";
+      allowedVersions = "^${lib.versions.majorMinor finalAttrs.version}.[0-9]+$";
+    };
```
Even if the allowed versions are detected correctly, the updater script tries to patch generic.nix instead of e.g. 3_12.nix. Any ideas on how to fix that?

flokli replied:
No idea, but if the update script doesn't work, it should be a draft.

felixsinger replied:
Sorry, I've just marked it as a draft. I thought it's good enough to mark it as ready for review and also to get some eyes on this issue.

felixsinger replied:
Fixed it by using nix-update-script. Everything works now.

felixsinger committed:
gerrit: Replace gitUpdater with nix-update-script

The gitUpdater does not allow configuring a custom path pointing to
the Nix file containing the version and hash strings, which will be
needed in a later commit. nix-update-script does. Use it.

Note that customizing the path is not needed in this commit, but making
a dedicated commit reduces the diff later.

The following command was used to verify that updating still works:

    nix-shell maintainers/scripts/update.nix --argstr package gerrit

Signed-off-by: Felix Singer <felixsinger@posteo.net>
